kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/DataTypeParser.java

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.defaults.internal;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+import io.delta.kernel.types.*;
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+
+/**
+ * Parses Delta data types based on the
+ * <a href="https://github.com/delta-io/delta/blob/master/PROTOCOL.md#primitive-types">
+ *     serialization rules </a> outlined in the Delta Protocol.
+ */
+public class DataTypeParser {
+
+    private DataTypeParser() {}
+
+    public static StructType parseSchema(JsonNode json) {
+        DataType parsedType = parseDataType(json);
+        if (parsedType instanceof StructType) {
+            return (StructType) parsedType;
+        } else {
+            throw new IllegalArgumentException(
+                String.format("Could not parse %s as a valid StructType", json));
```

Comment:
given `json` could be multi-line? in that is the case print it at the end of the message.

Reply:
`Could not parse the following JSON as a valid StructType:\n%s`